src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -3105,7 +3105,11 @@ internal override int CompareKeys(int index1, int index2)
                 if (next == null) return index1 - index2;
                 return next.CompareKeys(index1, index2);
             }
-            return descending ? -c : c;
+            // -c will result in a negative value for int.MaxValue (-int.MinValue == int.MinValue).
+            // Flipping keys earlier is more likely to trigger something strange in a comparer,
+            // particularly as it comes to the sort being stable.
+            // 0.CompareTo(c) is always negative for a positive c and positive for a negative.
+            return descending ? 0.CompareTo(c) : c;
```

Comment:
At the time of coding I preferred the fact that `0.CompareTo(c)` would also do the right thing if `c` was `0`. That shouldn't happen though, and your suggestion has one less branch (I'd expect the jitter to do the right thing in inlining `CompareTo` but not necessarily in removing the branch that considers the possibility of `c == 0`). I shall change that line to your suggestion.